interfaces/prompting/constraints.go

```diff
@@ -97,17 +129,333 @@ func (c *Constraints) Match(path string) (bool, error) {
+		return nil, err
+	}
+	if c.PathPattern == nil {
+		return nil, prompting_errors.NewInvalidPathPatternError("", "no path pattern")
+	}
+	availablePerms, ok := interfacePermissionsAvailable[iface]
+	if !ok {
+		return nil, prompting_errors.NewInvalidInterfaceError(iface, availableInterfaces())
+	}
+	if len(c.Permissions) == 0 {
+		return nil, prompting_errors.NewPermissionsListEmptyError(iface, availablePerms)
+	}
+	var invalidPerms []string
+	permissionMap := make(PermissionMap, len(c.Permissions))
+	for _, perm := range c.Permissions {
+		if !strutil.ListContains(availablePerms, perm) {
+			invalidPerms = append(invalidPerms, perm)
+			continue
+		}
+		permissionMap[perm] = &PermissionEntry{
+			Outcome:  outcome,
+			Lifespan: lifespan,
+			Duration: duration,
+		}
+	}
+	if len(invalidPerms) > 0 {
+		return nil, prompting_errors.NewInvalidPermissionsError(iface, invalidPerms, availablePerms)
+	}
+	constraints := &Constraints{
+		PathPattern: c.PathPattern,
+		Permissions: permissionMap,
+	}
+	return constraints, nil
+}
+
+// PatchConstraints hold information about the applicability of the modified
+// rule to particular paths or permissions. A request will be matched by the
+// constraints if the requested path is matched by the path pattern (according
+// to bash's globstar matching) and one or more requested permissions are
+// denied in the permission map, or all of the requested permissions are
+// allowed in the map. When snapd modifies the rule, it converts
+// PatchConstraints to RuleConstraints, using the rule's existing constraints
+// wherever a field is omitted.
+//
+// Any permissions which are omitted from the new permission map are left
+// unchanged from the existing rule. To remove an existing permission from the
+// rule, the permission should map to null.
+type PatchConstraints struct {
+	PathPattern *patterns.PathPattern `json:"path-pattern,omitempty"`
+	Permissions PermissionMap         `json:"permissions,omitempty"`
+}
+
+// PatchRuleConstraints validates the receiving PatchConstraints and uses the
+// existing rule constraints to construct a new RuleConstraints.
+//
+// If the path pattern or permissions fields are omitted, they are left
+// unchanged from the existing rule. If the permissions field is present in
+// the patch constraints, then any permissions which are omitted from the
+// patch constrants' permission map are left unchanged from the existing rule.
+// To remove an an existing permission from the rule, the permission should map
+// to null in the permission map of the patch constraints.
+//
+// The existing rule constraints should never be modified.
+func (c *PatchConstraints) PatchRuleConstraints(existing *RuleConstraints, iface string, currTime time.Time) (*RuleConstraints, error) {
+	ruleConstraints := &RuleConstraints{
+		PathPattern: c.PathPattern,
+	}
+	if c.PathPattern == nil {
+		ruleConstraints.PathPattern = existing.PathPattern
```

Comment:
Since we never modify the data in the `PathPattern` struct after it's been constructed, and this is simply a pointer to that heap-allocated struct, it should be fine to do this I think.

In particular, the path pattern was allocated independently of any constraints, so we won't prevent the old constraints from being garbage collected, if that's the concern.